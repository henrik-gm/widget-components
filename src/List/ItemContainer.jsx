import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './ItemContainer.scss';


class ItemContainer extends Component {

   constructor(props) {
      super(props);
      this.onResize = this.onResize.bind(this);
   }

   /*
    * Called after component rendering to DOM.
    */
   componentDidMount() {
      window.addEventListener('resize', this.onResize);
   }

   /*
    * Called before removing component.
    */
   componentWillUnmount() {
      window.removeEventListener('resize', this.onResize);
   }

   /*
    * Updates element width on window resize.
    */
   onResize() {
      if (this.width !== null && this.props.onWidth) {
         this.props.onWidth(this.width);
      }
   }

   /*
    * Sets component's DOM element.
    * @param {HTMLElement} el Component's DOM element
    */
   set element(el) {
      this.el = el;

      if (this.width !== null && this.props.onWidth) {
         this.props.onWidth(this.width);
      }
   }

   /*
    * Current element width
    * @returns {number|null}
    */
   get width() {
      return this.el ? this.el.offsetWidth : null;
   }

   /*
    * Renders element.
    * @returns {XML}
    */
   render() {
      return (
         <div
            className={[styles.item, this.props.selected ? styles.selected : ''].join(' ')}
            onClick={this.props.onClick}
            ref={el => (this.element = el)}
         >
            {this.props.children}
            <i className={['KambiWidget-primary-background-color', styles.border].join(' ')} />
         </div>
      );
   }

}

ItemContainer.propTypes = {
   /*
    * On click handler
    */
   onClick: PropTypes.func,

   /*
    * Is this item currently selected?
    */
   selected: PropTypes.bool,

   /*
    * Item contents
    */
   children: PropTypes.node,

   /*
    * Called when item width is known
    */
   onWidth: PropTypes.func
};

export default ItemContainer;
