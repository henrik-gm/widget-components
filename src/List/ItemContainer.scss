:local(.item) {
   cursor: pointer;
   position: relative;
   transition: background-color .5s;
   padding-bottom: 3px;
   width: 17.86%;
   height: 44px;
   align-items: center;
   display: flex;

   &:hover {
      background-color: #f0f0f0;
   }

   :local(.border) {
      position: absolute;
      bottom: 0;
      height: 3px;
      left: 0;
      right: 0;
      visibility: hidden;
   }

   &.selected {
      font-weight: bold;

      :local(.border) {
         visibility: visible;
      }
   }
}
