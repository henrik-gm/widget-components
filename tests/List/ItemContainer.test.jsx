/* eslint-env jest */
import React from 'react';
import ItemContainer from '../../src/List/ItemContainer';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import { shallow, mount } from 'enzyme';

let renderer;

describe('ItemContainer view', () => {

   beforeEach(() => {
      renderer = new ReactShallowRenderer();
   });

   it('renders correctly with default props', () => {
      expect(renderer.render(
         <ItemContainer />
      )).toMatchSnapshot();
   });

   it('renders correctly selected state', () => {
      expect(renderer.render(
         <ItemContainer selected={true} />
      )).toMatchSnapshot();
   });

   it('renders correctly with one child element', () => {
      expect(renderer.render(
         <ItemContainer>One</ItemContainer>
      )).toMatchSnapshot();
   });

   it('renders correctly with many child elements', () => {
      expect(renderer.render(
         <ItemContainer>
            One
            Two
            <ul>
               <li>Three</li>
            </ul>
         </ItemContainer>
      )).toMatchSnapshot();
   });

});

describe('ItemContainer interface', () => {

   it('handles click events correctly', () => {
      const onClickMock = jest.fn();

      const wrapper = shallow(
         <ItemContainer onClick={onClickMock} />
      );

      expect(onClickMock).not.toHaveBeenCalled();

      wrapper.first().simulate('click');

      expect(onClickMock).toHaveBeenCalledTimes(1);
   });

   it('calls onWidth after mounting', () => {
      const onWidthMock = jest.fn();

      mount(<ItemContainer onWidth={onWidthMock} />);

      expect(onWidthMock).toHaveBeenCalledTimes(1);
   });

   it('unmounts correctly', () => {
      const wrapper = mount(<ItemContainer />);
      wrapper.unmount();
   });


   it('calls onWidth on window resize', () => {
      const mockOnWidth = jest.fn();

      const wrapper = mount(<ItemContainer onWidth={mockOnWidth} />);

      expect(mockOnWidth).toHaveBeenCalledTimes(1);

      window.dispatchEvent(new Event('resize'));

      expect(mockOnWidth).toHaveBeenCalledTimes(2);

   });


   it('behaves correctly when onWidth is not set', () => {
      const wrapper = mount(<ItemContainer />);

      window.dispatchEvent(new Event('resize'));
   });


});
