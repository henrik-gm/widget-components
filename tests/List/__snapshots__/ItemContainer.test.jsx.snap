exports[`ItemContainer view renders correctly selected state 1`] = `
<div
  className="item selected"
  onClick={undefined}>
  <i
    className="KambiWidget-primary-background-color border" />
</div>
`;

exports[`ItemContainer view renders correctly with default props 1`] = `
<div
  className="item "
  onClick={undefined}>
  <i
    className="KambiWidget-primary-background-color border" />
</div>
`;

exports[`ItemContainer view renders correctly with many child elements 1`] = `
<div
  className="item "
  onClick={undefined}>
  One Two
  <ul>
    <li>
      Three
    </li>
  </ul>
  <i
    className="KambiWidget-primary-background-color border" />
</div>
`;

exports[`ItemContainer view renders correctly with one child element 1`] = `
<div
  className="item "
  onClick={undefined}>
  One
  <i
    className="KambiWidget-primary-background-color border" />
</div>
`;
